src/dawn/SIR/SIR.cpp

```diff
@@ -174,4 +174,137 @@ std::shared_ptr<sir::StencilCall> sir::StencilCall::clone() const {
+
+    if(Stencils.size() > 0) {
+      retval &= std::equal(Stencils.begin(), Stencils.end(), rhs.Stencils.begin(),
+                           sharedpointer_comparison<sir::Stencil>);
+    }
+
+    if(StencilFunctions.size() > 0) {
+      retval &=
+          std::equal(StencilFunctions.begin(), StencilFunctions.end(), rhs.StencilFunctions.begin(),
+                     sharedpointer_comparison<sir::StencilFunction>);
+    }
+
+    if(GlobalVariableMap != nullptr && rhs.GlobalVariableMap != nullptr) {
+      if(GlobalVariableMap->size() == rhs.GlobalVariableMap->size()) {
+        if(GlobalVariableMap->size() > 0) {
+          retval &=
+              pointerMapComparison(*(GlobalVariableMap.get()), *(rhs.GlobalVariableMap.get()));
+        }
+      } else {
+        return false;
+      }
+    } else {
+      return GlobalVariableMap == rhs.GlobalVariableMap;
+    }
+
+    return retval;
+  }
+  return false;
+}
+bool SIR::operator!=(const SIR& rhs) const { return !(*this == rhs); }
+
+bool sir::Stencil::operator==(const sir::Stencil& rhs) const {
+  bool retval = true;
+  if(Fields.size() != rhs.Fields.size()) {
+    return false;
+  }
+  retval &= (Name == rhs.Name);
+  retval &= (Attributes == rhs.Attributes);
+  if(StencilDescAst != nullptr && rhs.StencilDescAst != nullptr) {
+    if(StencilDescAst->getRoot() != nullptr && rhs.StencilDescAst->getRoot() != nullptr) {
+      retval &= StencilDescAst->getRoot().get()->equals(rhs.StencilDescAst->getRoot().get());
+    } else {
+      return StencilDescAst->getRoot() == rhs.StencilDescAst->getRoot();
+    }
+  } else {
+    return StencilDescAst == rhs.StencilDescAst;
+  }
+  if(Fields.size() > 0) {
+    retval &= std::equal(Fields.begin(), Fields.end(), rhs.Fields.begin(),
+                         sharedpointer_comparison<Field>);
+  }
+
+  return retval;
+}
+
+bool sir::StencilFunction::operator==(const sir::StencilFunction& rhs) const {
+  bool retval = true;
+  retval &= (Name == rhs.Name);
+  retval &= Attributes == rhs.Attributes;
+  if(Args.size() != rhs.Args.size() && Intervals.size() != rhs.Intervals.size() &&
+     Asts.size() != rhs.Asts.size()) {
+    return false;
+  } else {
+    if(Args.size() > 0) {
+      retval &= std::equal(Args.begin(), Args.end(), rhs.Args.begin(),
+                           sharedpointer_comparison<sir::StencilFunctionArg>);
+    }
+    if(Intervals.size() > 0) {
+      retval &= std::equal(Intervals.begin(), Intervals.end(), rhs.Intervals.begin(),
+                           sharedpointer_comparison<sir::Interval>);
+    }
+
+    if(Asts.size() > 0) {
+      auto astcomparison = [](std::shared_ptr<dawn::AST> comparate1,
+                              std::shared_ptr<dawn::AST> comparate2) {
+        if(comparate1->getRoot() != nullptr && comparate2->getRoot() != nullptr) {
+          return comparate1->getRoot().get()->equals(comparate2->getRoot().get());
+        } else {
+          return comparate1->getRoot() == comparate2->getRoot();
```

Comment:
You don't need this? It should be just `return false`

Reply:
why? two ASTs with empy roots are the same?

Reply:
Point taken.